src/python_testing/TC_DeviceBasicComposition.py

```diff
@@ -941,6 +941,217 @@ def log_structured_data(start_tag: str, dump_string):
+                                [Clusters.Descriptor][Clusters.Descriptor.Attributes.DeviceTypeList]]
+        cluster_revision = self.endpoints[0][Clusters.Descriptor][Clusters.Descriptor.Attributes.ClusterRevision]
+        self.print_step("1a.4", "Verify that PartsList attribute is not empty and contains a list of all endpoints.")
+        if len(parts_list_ep_0) == 0:
+            self.fail_current_test()
+        self.print_step("1a.5", "Verify that Revision should match the revision of that Device Type.")
+        if device_revision_ep_0[0] != 1:
+            self.fail_current_test()
+        self.print_step("1a.5", "Verify that cluster Revision should not be less than 1")
+        if cluster_revision < 1:
+            self.fail_current_test()
+        self.print_step(
+            "1a.2", "Verify that the DeviceTypeList count is at least one.")
+
+        self.print_step(
+            "1a.3", "Verify that the DeviceTypeList contains one Root Node Device Type.")
+
+        if ROOT_NODE_DEVICE_TYPE not in listed_device_types:
+            self.record_error(self.get_test_name(), location=AttributePathLocation(endpoint_id=0),
+                              problem="Root node device type not listed on endpoint 0", spec_location="Root node device type")
+
+            self.fail_current_test()
+        self.print_step(
+            "1a.3", "Verify that the DeviceTypeList may only contain other Node Device Types (device types with scope=node, it can be any of the following Power Source, OTA Requestor, OTA Provider) next to the Root Node Device Type.")
+
+        if len(listed_device_types) > 0:
+            if not all(x in {0x16, 0x0012, 0x0014} for x in listed_device_types):
+                self.fail_current_test()
+        self.print_step(3, "TH reads PartsList attribute- covered in step 2")
+
+        for endpoint_id, endpoint in self.endpoints.items():
+
+            self.print_step(8, "TH reads ServerList attribute for endpoint :{endpoint_id}")
+
+            if endpoint_id != 0:
+                if Clusters.Descriptor not in endpoint:
+                    continue
+
+                self.print_step(
+                    "1b", "TH reads DeviceTypeList and PartsList attributes from DUT for Endpoint {endpoint_id} supported by DUT (except Endpoint 0).")
+                parts_list_per_ep_non_zero = endpoint[Clusters.Descriptor][Clusters.Descriptor.Attributes.PartsList]
+
+                listed_device_types_ep_non_zero = [i.deviceType for i in endpoint
+                                                   [Clusters.Descriptor][Clusters.Descriptor.Attributes.DeviceTypeList]]
+                cluster_revision_ep_non_zero = endpoint[Clusters.Descriptor][Clusters.Descriptor.Attributes.ClusterRevision]
+
+                self.print_step(
+                    "1b.1", "Verify that the DeviceTypeList count is at least one for end point {endpoint_id}")
+
+                if len(listed_device_types_ep_non_zero) < 1:
```

Comment:
it is more pythonic to do:

```suggestion
                if not listed_device_types_ep_non_zero:
```